File: linked-list/142.js
// Tc = O(n)
// Sc = O(1)
// Floyd cycle detection algorithm
var detectCycle = function(head) {
    if (!head || !head.next) {
        return null;
    }

    let slow = head;
    let fast = head;

    // Phase 1: Detect Cycle
    while (fast !== null && fast.next !== null) {
        slow = slow.next;
        fast = fast.next.next;

        if (slow === fast) {
            break; // Cycle detected
        }
    }

    // If no cycle was found (fast reached the end of the list)
    if (slow !== fast) {
        return null;
    }

    // Phase 2: Find Cycle Start
    let p = head;
    while (p !== slow) {
        p = p.next;
        slow = slow.next;
    }

    return p;
};

File: linked-list/142.py
# Tc = O(n)
# Sc = O(1)
# Floyd's Cycle-Finding Algorithm
class ListNode:
    def __init__(self, val = 0, next = None):
        self.val = val
        self.next = next
class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return None

        slow = head
        fast = head

        # Phase 1: Detect Cycle
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                break # Cycle detected

        # If no cycle was found (fast reached the end of the list)
        if slow != fast:
            return None

        # Phase 2: Find Cycle Start
        p = head
        while p != slow:
            p = p.next
            slow = slow.next

        return p

